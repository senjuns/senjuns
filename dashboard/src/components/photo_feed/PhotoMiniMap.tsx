import Konva from 'konva';
import { FC, useRef } from 'react';
import { Image, Layer, Rect, Stage } from 'react-konva';
import useImage from 'use-image';

import { Size } from 'shared/interfaces';

interface PhotoMiniMapProps {
  /**
   * The current cursor position in the Main Window.
   */
  cursorPosition: Konva.Vector2d;
  /**
   * `true` if the minimap is draggable, `false` otherwise.
   */
  draggable?: boolean;
  /**
   * Size of the image to be displayed as a background image.
   */
  imageSize: Size;
  /**
   * The size of image to be displayed in minimap.
   */
  miniMapSize: Size;
  /**
   * The image url that will be used as a background of minimap.
   */
  miniMapImageUrl: string;
  /**
   * The size of the presentation layer.
   */
  presentationSize: Size;
  /**
   * The current zoom scale in Main Window.
   */
  scale: number;
  /**
   * Callback to be called when the position of minimap changes.
   */
  onChangeMiniMapPosition: (position: Konva.Vector2d) => void;
}

const PhotoMiniMap: FC<PhotoMiniMapProps> = ({
  cursorPosition,
  draggable = true,
  imageSize,
  miniMapSize,
  miniMapImageUrl,
  presentationSize,
  scale,
  onChangeMiniMapPosition,
}) => {
  const [image] = useImage(miniMapImageUrl);

  const stageRef = useRef<Konva.Stage>(null);
  const viewBoxRef = useRef<Konva.Rect>(null);
  const viewScaleX = miniMapSize.width / imageSize.width;
  const viewScaleY = miniMapSize.height / imageSize.height;
  const viewDimensions = {
    width: presentationSize.width / scale,
    height: presentationSize.height / scale,
  };
  const viewXPosition = -cursorPosition.x / scale;
  const viewYPosition = -cursorPosition.y / scale;

  const handleDragMove = () => {
    const viewBox = viewBoxRef.current;
    const stage = stageRef.current;
    if (!viewBox || !stage) return;

    const { x, y } = viewBox.getPosition();
    const scale = stage.scaleX();

    const newPosition = {
      x: Math.min(
        Math.max(0, x),
        miniMapSize.width / scale - viewDimensions.width
      ),
      y: Math.min(
        Math.max(0, y),
        miniMapSize.height / scale - viewDimensions.height
      ),
    };

    viewBox.position(newPosition);
    onChangeMiniMapPosition({ x: newPosition.x * -1, y: newPosition.y * -1 });
  };

  return (
    <Stage
      ref={stageRef}
      width={miniMapSize.width}
      height={miniMapSize.height}
      scaleX={viewScaleX}
      scaleY={viewScaleY}
      style={{
        width: miniMapSize.width,
      }}
    >
      <Layer id="viewbox">
        <Image
          image={image}
          height={miniMapSize.height / viewScaleY}
          width={miniMapSize.width / viewScaleX}
          x={0}
          y={0}
        />

        <Rect
          draggable={draggable}
          x={viewXPosition}
          y={viewYPosition}
          width={viewDimensions.width}
          height={viewDimensions.height}
          fill="#E86339"
          opacity={0.5}
          cornerRadius={80}
          ref={viewBoxRef}
          onDragMove={handleDragMove}
        />
      </Layer>
    </Stage>
  );
};

export default PhotoMiniMap;
