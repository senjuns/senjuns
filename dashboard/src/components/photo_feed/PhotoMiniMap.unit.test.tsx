import * as ShallowRenderer from 'react-test-renderer/shallow';
import PhotoMiniMap from './PhotoMiniMap';

describe('<PhotoMiniMap />', () => {
  it('renders correctly.', async () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <PhotoMiniMap
        cursorPosition={{ x: 0, y: 0 }}
        draggable={false}
        imageSize={{ height: 1000, width: 1000 }}
        miniMapSize={{ height: 100, width: 100 }}
        miniMapImageUrl=""
        presentationSize={{ height: 500, width: 500 }}
        scale={0}
        onChangeMiniMapPosition={jest.fn()}
      />
    );

    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
