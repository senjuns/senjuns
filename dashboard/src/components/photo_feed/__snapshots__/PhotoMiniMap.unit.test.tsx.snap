// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoMiniMap /> renders correctly. 1`] = `
<ForwardRef
  height={100}
  scaleX={0.1}
  scaleY={0.1}
  style={
    Object {
      "width": 100,
    }
  }
  width={100}
>
  <Layer
    id="viewbox"
  >
    <Image
      height={1000}
      width={1000}
      x={0}
      y={0}
    />
    <Rect
      cornerRadius={80}
      draggable={false}
      fill="#E86339"
      height={Infinity}
      onDragMove={[Function]}
      opacity={0.5}
      width={Infinity}
      x={NaN}
      y={NaN}
    />
  </Layer>
</ForwardRef>
`;
